libstorage/storage.c

```diff
@@ -387,110 +464,137 @@ int storage_unregisterfs(const char *name)
 }
 
 
-int storage_mountfs(storage_t *dev, const char *name, const char *data, unsigned long mode, oid_t *root)
+int storage_mountfs(storage_t *strg, const char *name, const char *data, unsigned long mode, oid_t *root)
 {
-	filesystem_t *fs = storage_getfs(name);
-	filesystem_ctx_t *ctx;
 	int err;
+	storage_fsctx_t *fsctx;
+	storage_fsHandler_t *handler = storage_getfs(name);
 
-	if ((dev == NULL) || (dev->ctx == NULL) || (dev->parts != NULL) || (fs == NULL) || (root == NULL))
+	if ((strg == NULL) || (strg->dev == NULL) || (strg->parts != NULL) || (handler == NULL) || (root == NULL))
 		return -EINVAL;
 
-	if (dev->fs != NULL)
+	if (strg->fs != NULL)
 		return -EBUSY;
 
-	if ((ctx = malloc(sizeof(filesystem_ctx_t))) == NULL)
+	fsctx = malloc(sizeof(storage_fsctx_t));
+	if (fsctx == NULL)
+		return -ENOMEM;
+
+	strg->fs = malloc(sizeof(storage_fs_t));
+	if (strg->fs == NULL) {
+		free(fsctx);
 		return -ENOMEM;
+	}
 
-	if ((err = storagectx_init(&ctx->ctx, fs->handler, NULL)) < 0) {
-		free(ctx);
+	err = storagectx_init(&fsctx->reqctx, storage_fsHandler);
+	if (err < 0) {
+		free(fsctx);
+		free(strg->fs);
```

Comment:
Imho it might be a good idea, you depend on this pointer being NULL. Applies to other exit paths too
```suggestion
		free(strg->fs);
		stg->fs = NULL;
```

Reply:
I agree with You.